containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcClient.java

```diff
@@ -0,0 +1,280 @@
+                "response=" + response.getResponse() + ", " +
+                "message.length=" + response.getMessage().length() + "}");
+            logFunction.apply("messages equal = " + message.equals(response.getMessage()));
+            return response;
+
+        } finally {
+            shutdownChannel(channel);
+        }
+    }
+
+    public static CourierSummary collectPackages(final String host,
+                                                 final int port,
+                                                 final Map<String, Object> headers,
+                                                 final String idPrefix,
+                                                 final String from,
+                                                 final List<String> messages,
+                                                 final int interMessageSleepMs,
+                                                 final boolean lockStepMode) throws InterruptedException {
+        final ManagedChannel channel = initializeChannel(host, port);
+
+        try {
+            final Semaphore lockStep = new Semaphore(1);
+
+            final Channel wrappedChannel = wrapResponseLogger(channel);
+            final Metadata headerMetadata = attachRequestHeaders(headers);
+
+            final CourierGrpc.CourierStub rawStub = CourierGrpc.newStub(wrappedChannel);
+            final CourierGrpc.CourierStub futureStub = MetadataUtils.attachHeaders(rawStub, headerMetadata);
+
+            logFunction.apply("will try to send package from " + from + " ...");
+
+            final CompletableFuture<CourierSummary> responsePromise = new CompletableFuture<>();
+            try {
+                final StreamObserver<CourierRequest> collector =
+                    futureStub.collectPackages(new StreamObserver<CourierSummary>() {
+
+                        private long numMessages = 0;
+                        private long totalLength = 0;
+
+                        @Override
+                        public void onNext(final CourierSummary response) {
+                            logFunction.apply("received response CourierSummary{" +
+                                "count=" + response.getNumMessages() + ", " +
+                                "length=" + response.getTotalLength() + "}");
+                            numMessages += response.getNumMessages();
+                            totalLength += response.getTotalLength();
+                            if (lockStepMode) {
+                                logFunction.apply("releasing semaphore after receiving response");
+                                lockStep.release();
+                            }
+                        }
+
+                        @Override
+                        public void onError(final Throwable throwable) {
+                            logFunction.apply("error in collecting summaries " + throwable);
+                            responsePromise.complete(null);
```

Comment:
Should we be somehow interrupting the client message loop if there's an error here?  Could it become deadlocked?